src/esphomelib/light/neo_pixel_bus_light_output.h

```diff
@@ -0,0 +1,211 @@
+      return;
+
+    uint32_t now = micros();
+    this->last_refresh_ = now;
+    this->next_show_ = false;
+
+    ESP_LOGVV(TAG, "Writing RGB values to bus...");
+
+#ifdef USE_OUTPUT
+    if (this->power_supply_ != nullptr) {
+      bool is_light_on = false;
+      for (int i = 0; i < this->controller_->PixelCount(); i++) {
+        auto color = this->controller_->GetPixelColor(i);
+        if (is_on(color)) {
+          is_light_on = true;
+          break;
+        }
+      }
+
+      if (is_light_on && !this->has_requested_high_power_) {
+        this->power_supply_->request_high_power();
+        this->has_requested_high_power_ = true;
+      }
+      if (!is_light_on && this->has_requested_high_power_) {
+        this->power_supply_->unrequest_high_power();
+        this->has_requested_high_power_ = false;
+      }
+    }
+#endif
+
+    this->controller_->Show();
+  }
+  float get_setup_priority() const override {
+    return setup_priority::HARDWARE;
+  }
+  template <typename U>
+  static typename std::enable_if<std::is_same<U, RgbColor>::value, RgbColor>::type
+  get_light_color(LightState *state, const LightColorValues values) {
+    float red, green, blue;
+    values.as_rgb(&red, &green, &blue);
+    red = gamma_correct(red, state->get_gamma_correct());
+    green = gamma_correct(green, state->get_gamma_correct());
+    blue = gamma_correct(blue, state->get_gamma_correct());
+    uint8_t redb = red * 255;
+    uint8_t greenb = green * 255;
+    uint8_t blueb = blue * 255;
+    return RgbColor(redb, greenb, blueb);
+  }
+
+  template <typename U>
+  static typename std::enable_if<std::is_same<U, RgbwColor>::value, RgbwColor>::type
+  get_light_color(LightState *state, const LightColorValues values) {
+    float red, green, blue, white, brightness;
+    values.as_rgbw(&red, &green, &blue, &white);
+    values.as_brightness(&brightness);
+    red = gamma_correct(red, state->get_gamma_correct());
+    green = gamma_correct(green, state->get_gamma_correct());
+    blue = gamma_correct(blue, state->get_gamma_correct());
+    white = gamma_correct(white, state->get_gamma_correct());
+    uint8_t redb = red * 255;
+    uint8_t greenb = green * 255;
+    uint8_t blueb = blue * 255;
+    uint8_t whiteb = white * 255;
+    uint8_t brightnessb = brightness * 255;
+    // currently in hass there is no way to only show white via the hass ui, so disable the colors if all of them are on and therefore very white
+    bool white_colors = redb >= brightnessb - 10 && greenb >= brightnessb - 10 && blueb >= brightnessb - 10 && whiteb >= brightnessb - 10;
+    if (white_colors) {
+      return RgbwColor(whiteb);
+    } else {
+      return RgbwColor(redb, greenb, blueb, whiteb);
+    }
+  }
+
+ protected:
+  NeoPixelBus<T_COLOR_FEATURE, T_METHOD> *controller_{nullptr};
+  uint32_t last_refresh_{0};
+  bool next_show_{true};
+#ifdef USE_OUTPUT
+  PowerSupplyComponent *power_supply_{nullptr};
+  bool has_requested_high_power_{false};
+#endif
+ private:

```

Comment:
When writing code for other languages/libraries, I would agree. In fact, up until esphomelib 1.3 or so everything was declared private.

But ... around that time I started adding lots of new integrations by using existing libraries. Sometimes though, I *needed* to use an internal attribute that the author of the library didn't intend to be used. So then I had to fork the library each time I needed the tiniest of changes (ended up being 5 or so forks, which is just a terrible mess to maintain).

So the rationale is: If someone already goes to the trouble of extending this class (which wouldn't be required otherwise), why not give them access to this. I'm sure people can figure out weird use-cases where these hacks are somehow necessary. And as long as it works for them and they know what they're doing, I don't care.

One of the main reasons of using `private` visibility is for stopping people from using undocumented APIs. If lots of people end up using the undocumented API, it effectively becomes an API that has to be maintained, as many people are depending on it not breaking. However, esphomelib has tons of API breaking changes each release anyway, and the only thing that I try to create migrations for is YAML configurations.

So, I don't care if people use these attributes, as they hopefully know what they're doing.